### PR TITLE
Add a `test` command

### PR DESCRIPTION
**Pull request by jakemac53**
This is ugly under the covers - it does roughly this:

- runs a build
- runs the create_merged_dir command
- creates missing html files for package:test manually
- runs pub run test with --precompiled

However, this gets things working with a mix of custom html files and not.

Unfortunately this does break running tests with --pub-serve if you don't have custom html files written (see the skipped test), but I think that is fine.

### CODE REVIEW

File: build_barback/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+## 0.5.0+1
+
+- Stop casting `BuildStep` to `BuildStepImpl` and remove `src` import from
```
Comment (natebosch) [inline build_barback/CHANGELOG.md]: 👍

File: build_runner/lib/src/entrypoint/options.dart [hunk 2]
```diff
@@ -200,3 +209,106 @@ class _ServeCommand extends _WatchCommand {
...
+  }
+
+  /// Creates a merged directory for the current build script.
+  Future<String> _createMergedDir() async {
+    var tmpDir = await Directory.systemTemp.createTemp('build_runner_test');
+    var tmpDirPath = tmpDir.absolute.uri.toFilePath();
+    var scriptLocation = Platform.script.path;
+    var process = await new ProcessManager().spawn('pub', [
+      'run',
+      'build_runner:create_merged_dir',
```
Comment (natebosch) [inline build_runner/lib/src/entrypoint/options.dart]: can you file an issue to make this a better supported thing?

Comment (natebosch) [reply build_runner/lib/src/entrypoint/options.dart]: I guess https://github.com/dart-lang/build/issues/387 covers it :)

**Comment by jakemac53** [comment]
looks like there is some minor issue happening on windows - will likely investigate that after the break

**Pull request merged**